library/cipher_wrap.c

```diff
@@ -2277,4 +2295,92 @@ const mbedtls_cipher_definition_t mbedtls_cipher_definitions[] =
                      sizeof(mbedtls_cipher_definitions[0]))
 int mbedtls_cipher_supported[NUM_CIPHERS];
 
+const mbedtls_cipher_base_t *mbedtls_cipher_base_lookup_table[] = {
```

Comment:
Since the positions in this table must match the values of the `MBEDTLS_CIPHER_BASE_INDEX_xxx` defines, is there any way to tie the two together so that they can't get out of sync? e.g. having an enum for `MBEDTLS_CIPHER_BASE_INDEX_xxx` (so the compiler can ensure the values are sequential) and using designated initialisers for `mbedtls_cipher_base_lookup_table[]`?

Reply:
Maybe something like (untested)

```
enum MBEDTLS_CIPHER_BASE_INDEX {
    MBEDTLS_CIPHER_BASE_INDEX_EMPTY = 0,
#if defined(MBEDTLS_AES_C)
    MBEDTLS_CIPHER_BASE_INDEX_AES,
#endif
#if defined(MBEDTLS_ARIA_C)
    MBEDTLS_CIPHER_BASE_INDEX_ARIA,
#endif
#if defined(MBEDTLS_CAMELLIA_C)
    MBEDTLS_CIPHER_BASE_INDEX_CAMELLIA,
#endif

// etc...

#if defined(MBEDTLS_CIPHER_MODE_XTS) && defined(MBEDTLS_AES_C)
    MBEDTLS_CIPHER_BASE_INDEX_XTS_AES,
#endif
    MBEDTLS_CIPHER_BASE_INDEX_COUNT
};

const mbedtls_cipher_base_t *mbedtls_cipher_base_lookup_table[MBEDTLS_CIPHER_BASE_INDEX_COUNT] = {
    NULL,
#if defined(MBEDTLS_AES_C)
    [MBEDTLS_CIPHER_BASE_INDEX_AES] = &aes_info,
#endif
#if defined(MBEDTLS_ARIA_C)
    [MBEDTLS_CIPHER_BASE_INDEX_ARIA] = &aria_info,
#endif
#if defined(MBEDTLS_CAMELLIA_C)
    [MBEDTLS_CIPHER_BASE_INDEX_CAMELLIA] = &camellia_info,
#endif

// etc...

#if defined(MBEDTLS_CIPHER_MODE_XTS) && defined(MBEDTLS_AES_C)
    [MBEDTLS_CIPHER_BASE_INDEX_XTS_AES] = &xts_aes_info,
#endif
};
```

Reply:
This is great, I couldn't think of a good way to exclude not-used entries from the look-up table, but having an enum is a nice solution.

Reply:
No guarantee it will work with not-quite-really-c99-compliant compilers such as MSVC, of course (some compilers don't like initialising with constants and you have to fall back to macros; hence "untested") - but would be great if it does work!

Reply:
Well, CI was happy so I think it's OK